pkg/pillar/cmd/zedrouter/patchenvelopes.go

```diff
@@ -0,0 +1,257 @@
+// VolumeStatusCh and PatchEnvelopeInfoCh to process volumes and patch envelopes.
+// NewPatchEnvelopes() starts coroutine processMessages which reads from both channels
+// and stores latest state of patch envelopes exposed to app instances via Get()
+// Note that composing envelopes (matching VolumeStatuses with PatchEnvelopeInfo) can take
+// significant time on big files with small CPU, therefore in order not to block writing
+// to channels, composing is happening on copy of PatchEnvelopeInfo and VolumeStatuses
+//
+// That means that VolumeStatusCh and PatchEnvelopeInfoCh should not be read from anywhere
+// except processMessages(). so there is no deadlocks.
+// except processMessages(), also it will not wait/hang by sending information to other channels
+// so there are no deadlocks.
+type PatchEnvelopes struct {
+	sync.Mutex
+	VolumeStatusCh      chan PatchEnvelopesVsCh
+	PatchEnvelopeInfoCh chan []types.PatchEnvelopeInfo
+	Wg                  sync.WaitGroup
+
+	composedEnvelopes []types.PatchEnvelopeInfo
+
+	envelopes         []types.PatchEnvelopeInfo
+	completedVolumes  []types.VolumeStatus
+	log               *base.LogObject
+	latestVolumeIndex int
+}
+
+// PatchEnvelopesVsChAction specifies action for patch envelope volume status
+type PatchEnvelopesVsChAction uint8
+
+const (
+	// PatchEnvelopesVsChActionDelete -- delete reference with volume
+	PatchEnvelopesVsChActionDelete PatchEnvelopesVsChAction = iota
+	// PatchEnvelopesVsChActionPut -- create or update reference with volume
+	PatchEnvelopesVsChActionPut
+)
+
+// PatchEnvelopesVsCh is a wrapper for VolumeStatus specifying action
+// for external Patch Envelope
+type PatchEnvelopesVsCh struct {
+	Vs     types.VolumeStatus
+	Action PatchEnvelopesVsChAction
+}
+
+// NewPatchEnvelopes returns PatchEnvelopes structure and starts coroutine
+// to process messages from channels. Note that we create buffered channels
+// to avoid unbounded processing time in writing to channel
+func NewPatchEnvelopes(log *base.LogObject) *PatchEnvelopes {
+	pe := &PatchEnvelopes{
+		VolumeStatusCh:      make(chan PatchEnvelopesVsCh, 1),
+		PatchEnvelopeInfoCh: make(chan []types.PatchEnvelopeInfo, 1),
+
+		log: log,
+	}
+
+	go pe.processMessages()
+
+	return pe
+}
+
+// Get returns list of Patch Envelopes available for this app instance
+func (pes *PatchEnvelopes) Get(appUUID string) types.PatchEnvelopeInfoList {
+	pes.Lock()
+
+	newMessages := pes.patchEnvelopesChanged()
+	currentVs := make([]types.VolumeStatus, len(pes.completedVolumes))
+	currentPeInfo := make([]types.PatchEnvelopeInfo, len(pes.envelopes))
+	currentVolumeIndex := pes.latestVolumeIndex
+
+	if newMessages {
+		copy(currentVs, pes.completedVolumes)
+		copy(currentPeInfo, pes.envelopes)
+		pes.latestVolumeIndex = len(pes.completedVolumes)
+	}
+
+	// composing envelopes might take a lot of time, in order to prevent channels
+	// from being blocked we work on copy of data so that new changes can be
+	// processed
```

Comment:
But the channel will still be blocked until pe.processMessages() gets back to reading the channel which can take time if you're computing a sha256 over the file. Hmm - but for a downloaded file we should already have the sha256 in VolumeStatus I think. Can you check.

In any case, the mutex has nothing to do with how the channel is read AFAIU.

Reply:
We have independent goroutine which calls `pes.processMessages()` which only writes to `pes.envelopes` and `pes.completedVolumes` we condition this writing on mutex since we might be copying. Zedrouter during handlers might call `pes.Get()` which will block mutex until it'll copy pes.envelopes and pes.completedVolumes and then will start to compute files sha if needed. So nothing is blocking channel from writing to it except some copying. But if we have sha256 chached in VolumeStatus it'll make this faster, will check

Reply:
`blobFromVolRef` is only used in `pes.Get()` now

Reply:
But that's just transforming the problem to a different problem.
Suppose an app instance polls the API endpoint every 10 seconds and it takes 2 minutes to compute the sha on the large blob. 
Q: How many goroutines will be running computing the sha over the same file?
A: infinite

Reply:
@eriknordmark Now we take SHA from `ContentTreeStatus` and determine which `PatchEnvelopes` require changes in handlers and then send notification to `PatchEnvelope` to update specific envelope. I tested with controller and it works.

Reply:
OK so the sha is taken care of.